### PR TITLE
Fix RT-DETR cache for generate_anchors

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

For RT-DETR model:

1. Fix lru_cache for the `generate_anchors` method, so even if we go with dynamic anchors generation we only generate it once per image of the same size.
2. Fix anchors dtype: make anchors generated always in float32, and then convert to the desired type. Otherwise, a difference might be observed between dynamic and static anchors inference in float16/bfloat16 (test is added).

## Who can review?
cc @SangbumChoi

### CODE REVIEW

**Review by SangbumChoi — APPROVED** [review]
Thanks for handling this!
(Did you also finetuned RT-DETR with bfloat16? Is it okay?)

**Comment by qubvel** [comment]
> Did you also finetuned RT-DETR with bfloat16? Is it okay?

No, I didn't. I tried bfloat16 only for inference 🙂

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

File: src/transformers/models/rt_detr/modeling_rt_detr.py [hunk 1]
```diff
@@ -1656,7 +1656,10 @@ def unfreeze_backbone(self):
             param.requires_grad_(True)
 
     @lru_cache(maxsize=32)
-    def generate_anchors(self, spatial_shapes=None, grid_size=0.05, dtype=torch.float32, device="cpu"):
+    def generate_anchors(self, spatial_shapes=None, grid_size=0.05):
+        # We always generate anchors in float32 to preserve original model code
```
Comment (qubvel) [inline src/transformers/models/rt_detr/modeling_rt_detr.py:1660]:
```suggestion
        # We always generate anchors in float32 to preserve equivalence between
        # dynamic and static anchor inference
```

**Pull request merged**